File: computer_science/data-structure/queue/queue.js
export class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

export class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.length = 0;
  }
  enqueue(value) {
    const node = new Node(value);
    if (!this.last && this.length == 0) {
      this.first = node;
      this.last = node;
    } else {
      this.last.next = node;
      this.last = node;
    }
    this.length++;
    return this;
  }

  dequeue() {
    if (!this.first) return undefined;
    const temp = this.first;
    this.first = temp.next;
    if (this.first === null) {
      this.last = null;
    }
    temp.next = null;
    this.length--;
    return temp;
  }
}

File: computer_science/data-structure/queue/queue.py
class Queue:
    def __init__(self):
        self.elements = {}
        self.head = 0
        self.tails = 0

    def enqueue(self, value):
        self.elements[self.tails] = value
        self.tails = self.tails+1

    def dequeue(self):
        if self.head > 0:
            backoup = self.elements[self.head]
            del self.elements[self.head]
            return backoup
        return -1

    def peek(self):
        return self.elements[self.head]

    def length(self):
        return self.tails - self.head

    def isEmpty(self):
        return self.tails - self.head
